src/core/client.ts

```diff
@@ -227,57 +227,67 @@ class ApiClient<R extends RawApi> {
-
-            const res = await new Promise<ApiResponse<ApiCallResult<M, R>>>(
-                (resolve, reject) => {
-                    const onHttpError = toHttpError(method, sensLogs, reject);
-                    function onOtherError(err: unknown) {
-                        abort();
-                        reject(err);
-                    }
-                    const config = formDataRequired
-                        ? createFormDataPayload(p, onOtherError)
-                        : createJsonPayload(p);
-                    const opts = {
-                        ...this.options.baseFetchConfig,
-                        signal: abortController.signal,
-                        ...config,
-                    };
-                    const timeoutHandle = setTimeout(() => {
-                        const msg =
-                            `Request to '${method}' timed out after 500 seconds`;
-                        onOtherError(new Error(msg));
-                    }, 500_000); // 500 seconds timeout, same as in Bot API
-                    fetch(url, opts).then((res) => res.json())
-                        .then(resolve)
-                        .catch(onHttpError)
-                        .finally(() => clearTimeout(timeoutHandle));
-                },
-            );
-            return res;
         }
+        // Regular request
+        const p = payload ?? {};
+        const controller = new AbortController();
+        const abort = combineAborts(controller, signal);
+        const options = { ...opts.baseFetchConfig, signal: controller.signal };
+        return await new Promise<ApiResponse<ApiCallResult<M, R>>>(
+            (resolve, reject) => {
+                let rejected = false; // prevents repeated rejections
+                /** Error handler for `fetch` */
+                function onHttpError(err: unknown) {
+                    if (rejected) return;
+                    rejected = true;
+                    let msg = `Network request for '${method}' failed!`;
+                    if (isTelegramError(err)) {
+                        msg += ` (${err.status}: ${err.statusText})`;
+                    }
+                    if (opts.sensitiveLogs && err instanceof Error) {
+                        msg += ` ${err.message}`;
+                    }
+                    reject(new HttpError(msg, err));
+                }
+                /** Error handler for timeouts and body stream errors */
+                function onOtherError(err: unknown) {
+                    if (rejected) return;
+                    rejected = true;
+                    reject(err);
+                    abort();
+                }
+                // Create fetch config
+                const config = formDataRequired
+                    ? createFormDataPayload(p, onOtherError)
+                    : createJsonPayload(p);
+                // Use same timeouts for requests as Bot API
```

Comment:
```suggestion
                // Use same timeouts for requests as Bot API
                if (rejected) return; // don't start the request if createFormDataPayload already rejected
```

Reply:
You're technically right if we regard the payload creation as a black box. Practically, this can never happen. The error callback can only be invoked when the body is consumed, which only happens as soon as it's passed to `fetch`

Reply:
oh, right, i missed that